File: leetcode/medium/Q114_FlattenBinaryTreetoLinkedList.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def flatten(self, root: TreeNode) -> None:
        if root:
            self.refactor(root)

    def refactor(self, node: TreeNode) -> TreeNode:
        if node.left is None and node.right is None:
            return node

        left, re_left = node.left, node.left
        right = node.right

        if right:
            self.refactor(node.right)

        if left:
            re_left = self.refactor(node.left)
            node.right = left
            re_left.right = right
            node.left = None

        ret = node.right
        while ret.right:
            ret = ret.right
        return ret

File: src/medium/Q114_FlattenBinaryTreetoLinkedList.java
package medium;

public class Q114_FlattenBinaryTreetoLinkedList {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
          this.val = val;
          this.left = left;
          this.right = right;
        }
    }
     
    class Solution {
        public void flatten(TreeNode root) {
            reBalance(root);
        }
        public TreeNode reBalance(TreeNode node){
            if(node == null) return node;
            TreeNode leftEnd = null;
            if(node.left != null){
                leftEnd = reBalance(node.left);
            }
            if(node.right != null){
                reBalance(node.right);
            }
            if(leftEnd != null){
                TreeNode temp = node.right;
                node.right = node.left;
                leftEnd.right = temp;
                node.left = null;
            }
            while (node.right != null){
                node = node.right;
            }
            return node;
        }
    }
}
